File: src/classes/Pato.ts
import Comportamento from '../interfaces/Comportamento';
import PadraoCorredores from '../interfaces/PadraoCorredores';
import PadraoVoaveis from '../interfaces/PadraoVoaveis'

abstract class Pato {
	
	protected _comportamento_pato?: PadraoVoaveis
	
	protected _comportamento_pato_correr?: PadraoCorredores

	abstract mostrar(): string;
	
	public nadar(): string
	{
		return "Pato Nadando.";		
	}	

	public correr(): string {
		return "Não consigo correr...Que"
	}
	
	public disparar(): string {
		return "Se eu não corro, imagine disparar."
	}

	public setComportamento( padrao: PadraoVoaveis ): void
	{
		this._comportamento_pato = padrao;		
	}

	public setComportamentoCorrer(padrao: PadraoCorredores) : void {
		this._comportamento_pato_correr = padrao
	}
	
	public comportamento_pato(): string
	{
		return this._comportamento_pato?.voar() || "Sem comportamento de voar"
	}

	public comportamento_pato_correr(): string {
		return this._comportamento_pato_correr?.correr() ?? "Sem comportamento de correr"	
	}

	public comportamento_pato_disparar(): string {
		return this._comportamento_pato_correr?.disparar() ?? "Sem comportamento de dispato"
	}
}

export default Pato

File: EngSoft.Jogo.Patos/src/engsoft/jogo/patos/Pato.java
package engsoft.jogo.patos;

// Classe abstrata, ou seja, que contém pelo menos um método abstrato
public abstract class Pato {
	
	// Compisição através do atributo do tipo interface Padrao_Voaveis 
	protected Padrao_Voaveis comportamento_pato;

	// Em uma classe abstrata deve ter pelo menos um método abstrato, nesse caso é o método mostrar()
	abstract String mostrar();
	
	public String nadar()
	{
		return "Pato Nadando.";		
	}	
	
	// Este método garante delegação ao projeto, modificando comportamentos sem a necessidade de executa-lo novamente (tempo de compilação)
	public void setComportamento(Padrao_Voaveis padrao)
	{
		comportamento_pato = padrao;
	}
	
	// Implementação do atributo comportamento_pato
	public String comportamento_pato()
	{		
		return comportamento_pato.voar();		
	}
}
